airflow/providers/microsoft/azure/hooks/azure_data_factory.py

```diff
@@ -0,0 +1,659 @@
+        )
+
+    @provide_targeted_factory
+    def delete_linked_service(
+        self,
+        linked_service_name: str,
+        resource_group_name: Optional[str] = None,
+        factory_name: Optional[str] = None,
+        **config: Any,
+    ) -> None:
+        """
+        Delete the linked service:
+
+        :param linked_service_name: The linked service name.
+        :param resource_group_name: The linked service name.
+        :param factory_name: The factory name.
+        :param config: Extra parameters for the ADF client.
+        """
+
+        self.get_conn().linked_services.delete(
+            resource_group_name, factory_name, linked_service_name, **config
+        )
+
+    @provide_targeted_factory
+    def get_dataset(
+        self,
+        dataset_name: str,
+        resource_group_name: Optional[str] = None,
+        factory_name: Optional[str] = None,
+        **config: Any,
+    ) -> DatasetResource:
+        """
+        Get the dataset.
+
+        :param dataset_name: The dataset name.
+        :param resource_group_name: The resource group name.
+        :param factory_name: The factory name.
+        :param config: Extra parameters for the ADF client.
+        :return: The dataset.
+        """
+
+        return self.get_conn().datasets.get(resource_group_name, factory_name, dataset_name, **config)
+
+    @provide_targeted_factory
+    def update_dataset(
+        self,
+        dataset_name: str,
+        dataset: Dataset,
+        resource_group_name: Optional[str] = None,
+        factory_name: Optional[str] = None,
+        **config: Any,
+    ) -> DatasetResource:
+        """
+        Update the dataset.
+
+        :param dataset_name: The dataset name.
+        :param dataset: The dataset resource definition.
+        :param resource_group_name: The resource group name.
+        :param factory_name: The factory name.
+        :param config: Extra parameters for the ADF client.
+        :return: The dataset.
+        """
+
+        return self.get_conn().datasets.create_or_update(
+            resource_group_name, factory_name, dataset_name, dataset, **config
+        )
+
+    @provide_targeted_factory
+    def create_dataset(
+        self,
+        dataset_name: str,
+        dataset: Dataset,
+        resource_group_name: Optional[str] = None,
+        factory_name: Optional[str] = None,
+        **config: Any,
+    ) -> DatasetResource:
+        """
+        Create the dataset.
+
+        :param dataset_name: The dataset name.
+        :param dataset: The dataset resource definition.
+        :param resource_group_name: The resource group name.
+        :param factory_name: The factory name.
+        :param config: Extra parameters for the ADF client.
+        :return: The dataset.
+        """
+
+        return self.get_conn().datasets.create_or_update(
+            resource_group_name, factory_name, dataset_name, dataset, **config
+        )
```

Comment:
If both does the same shouldn't we only have one method, too? like `create_or_update`

Reply:
There are many `create_or_update`s. Same goes for them.

Reply:
I did it that way so that we can potentially add existence checks to the `create` methods and throw an exception when the entity already exists. But that could also be left to the user code.

Reply:
I see. Can you add this into the method instead?

Reply:
I had something like this in mind for the checks:

```python
@provide_targeted_factory
def create_factory(
    self,
    factory: Factory,
    resource_group_name: Optional[str] = None,
    factory_name: Optional[str] = None,
    **config: Any,
) -> Factory:
    """
    Create the factory.

    :param factory: The factory resource definition.
    :param resource_group_name: The resource group name.
    :param factory_name: The factory name.
    :param config: Extra parameters for the ADF client.
    :return: The factory.
    """

    factories = self.get_conn().factories

    try:
        factories.get(resource_group_name, factory_name)
    except CloudError as e:
        raise AirflowException(f"Factory {factory_name!r} already exists") from e
    else:
        return factories.create_or_update(resource_group_name, factory_name, factory, **config)
```

but this `CloudError` is a little too generic to my taste. What would you do ?

Reply:
Nice! I didn't know `raise from`. Looks really good IMO.

If that also is to generic you can also check for a kind of code or message?!
Sth like.
```python
    except CloudError as e:
        if e.status_or_code_or_message == 1234:
            raise AirflowException(f"Factory {factory_name!r} already exists") from e
        raise
```
WDYT?

Reply:
I'm glad you learned something from me 😋

Unfortunately, when you try to access something that doesn't exist it fails with a "403 - Unauthorized":

```
The client 'xxxx' with object id 'xxxx' does not have authorization to perform action 'Microsoft.DataFactory/factories/read' over scope '/subscriptions/xxxx/resourceGroups/xxxx/providers/Microsoft.DataFactory/factories/xxxx' or the scope is invalid. If access was recently granted, please refresh your credentials.
```

This could also be the case if the user doesn't set his credentials correctly. 😕